### PR TITLE
fix: fix import ERC20 on network 1337

### PR DESCRIPTION
**Pull request by sahar-fehri**
## Explanation

This PR fixes importing an ERC20 token on chainId 1337.
This issue is on extension and Mobile.

When you run a local ganache node, you use testDapp to deploy a token, then you try to add the token to your wallet; and go back to home page but you wont be able to see the token.

The error starts on this [line](https://github.com/MetaMask/core/blob/f8437e9a626007de7192871452273ebd0f2edb35/packages/assets-controllers/src/TokensController.ts#L335). It looks like it fails to pass that line to create the `newEntry` and push it to `newTokens` array. 
The error is thrown on this [line](https://github.com/MetaMask/core/blob/f8437e9a626007de7192871452273ebd0f2edb35/packages/assets-controllers/src/token-service.ts#L136) (See screenshot)

Screenshot 2024-01-12 at 18 14 32

I see that [tokenApi](https://github.com/consensys-vertical-apps/va-mmcx-token-api) does not support chainId 1337 and the error is being thrown here https://github.com/consensys-vertical-apps/va-mmcx-token-api/blob/6c1b749c33a110f9dd8004dd283f6d9f7dcb4824/src/server.ts#L522

that is because [networkConfig](https://github.com/consensys-vertical-apps/va-mmcx-token-api/blob/6c1b749c33a110f9dd8004dd283f6d9f7dcb4824/src/constants.ts#L399) does not contain the 1337 chainId. 

After testing also with Goerli testnet, where importing ERC20 works fine,  i noticed that goerli is not supported on core so the fetchTokenMetadata threw [here](https://github.com/MetaMask/core/blob/f8437e9a626007de7192871452273ebd0f2edb35/packages/assets-controllers/src/token-service.ts#L82)  because `isTokenDetectionSupportedForNetwork` returned false https://github.com/MetaMask/core/blob/f8437e9a626007de7192871452273ebd0f2edb35/packages/assets-controllers/src/assetsUtil.ts#L156. I wasnt sure why we did not treat chainId 1337 same way as goerli.
So i removed the `|| chainId === GANACHE_CHAIN_ID` check on the utils fct.

After the fix, i tested importing the ERC20, importing ERC721/1155, works fine.

Lemme know if anything!  🙏 

## Video Before

https://github.com/MetaMask/core/assets/10994169/7b567366-f5df-4d1b-b61d-7d4a46711faa

## Video After

https://github.com/MetaMask/core/assets/10994169/3068a526-4c95-4c88-9c62-38acb202561f

## References

* Fixes https://github.com/MetaMask/metamask-mobile/issues/6410

### `@metamask/assets-controllers`

- **<FIXED>**: Updates the fct on assetsUtils `isTokenListSupportedForNetwork` and removed the check for `GANACHE_CHAIN_ID`

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Review by mikesposito — COMMENTED** [review]
Changes look good, but I also do not have enough context on why `GANACHE_CHAIN_ID` is treated as an exception: perhaps it is a residual of what we had before the CoinGecko issue, or it is useful for e2e tests?

Were you able to run e2e tests on extension, using this PR's version of assets controllers?

**Comment by sahar-fehri** [comment]
> Changes look good, but I also do not have enough context on why `GANACHE_CHAIN_ID` is treated as an exception: perhaps it is a residual of what we had before the CoinGecko issue, or it is useful for e2e tests?
> 
> Were you able to run e2e tests on extension, using this PR's version of assets controllers?

Hey @mikesposito ! I made a small update so my change does modify the `isTokenListSupportedForNetwork` fct cause it seems to be also used in TokensListController, noticed that the test  `test/e2e/tests/add-hide-token.spec.js` was failing, but it passes now with latest update 🙏

File: packages/assets-controllers/src/token-service.ts [hunk 1]
```diff
@@ -78,7 +82,10 @@ export async function fetchTokenMetadata<T>(
   abortSignal: AbortSignal,
   { timeout = defaultTimeout } = {},
 ): Promise<T | undefined> {
-  if (!isTokenListSupportedForNetwork(chainId)) {
+  if (
+    !isTokenListSupportedForNetwork(chainId) ||
```
Comment (mcmire) [inline packages/assets-controllers/src/token-service.ts:86]: Sorry, I don't remember what was here before because you squashed the commit. If the goal of this function is to determine whether a chain supports token lists, what do you think about modifying the `isTokenListSupportedForNetwork` function to exclude the `GANACHE_CHAIN_ID`? Or did it cause the e2e test to fail as you mentioned? If that's the case, is there something we can do on the extension side to fix the test rather than making this code less cohesive?

Comment (sahar-fehri) [reply packages/assets-controllers/src/token-service.ts]: I have made an update so the fct `isTokenListSupportedForNetwork` does not consider the GanacheID as supported; (since the call inside fetchTokenMetadata to tokenApi would fail because token api does not support it anyway)

but also so that does not stop the test from doing the startPolling..

**Comment by mcmire** [comment]
Hi @sahar-fehri. Sorry it's taken me a while to look at this.

I took some time to test the original issue and look at the code. I believe I found exactly what you found, but just for clarity, this seems to be the sequence of events:

- The user adds the token to their wallet through a dapp, which calls `wallet_watchAsset` on the provider and to MetaMask.
- This causes the `addToken` method on TokensController to get called.
- This method calls `fetchTokenMetadata`, also on TokensController.
- This method calls `fetchTokenMetadata` in `assetsUtil`.
- This function attempts to request information about the token from the Token API. However, because the Token API doesn't support chain 1337, the request fails and an error is thrown.
- `fetchTokenMetadata` may throw an error if `isTokenListSupportedForNetwork` returns false. If this were to happen, then the TokensController version of `fetchTokenMetadata` would catch this error and return `undefined`. `addToken` expects this to happen and continues on without issue.

As for the failing extension e2e test — ["Add existing token using search renders the balance for the chosen token"](https://github.com/MetaMask/metamask-extension/blob/642f6d4cf1655edc519de31075d8b14abe0f8e83/test/e2e/tests/add-hide-token.spec.js#L83) — I looked at why this was failing too. This test seems to be trying to simulate production, because if you do remove `GANACHE_CHAIN_ID` from `isTokenListSupportedForNetwork`, then it removes the option to search for tokens from the UI. The way that the extension has gotten around this is with a check in [`import-tokens-modal.js`](https://github.com/MetaMask/metamask-extension/blob/642f6d4cf1655edc519de31075d8b14abe0f8e83/ui/components/multichain/import-tokens-modal/import-tokens-modal.js#L119), which turns it on in e2e tests.

I see why we make an exception for e2e tests here — we need _some_ way of testing the search box — but I wonder if the way to do this is to use a custom chain ID for Ganache (_not_ 1337, in other words) so that `isTokenListSupportedForNetwork` returns true. Perhaps we will need to extract this test to a different file so that this is easier, I'm not sure.

To recap, I think removing the `GANACHE_CHAIN_ID` check from `isTokenListSupportedForNetwork` seems to be the correct direction, and I also think that we might want to update the e2e test so that it better simulates production instead of localhost.

What do you think?

**Comment by sahar-fehri** [comment]
> Hi @sahar-fehri. Sorry it's taken me a while to look at this.
> 
> I took some time to test the original issue and look at the code. I believe I found exactly what you found, but just for clarity, this seems to be the sequence of events:
> 
> * The user adds the token to their wallet through a dapp, which calls `wallet_watchAsset` on the provider and to MetaMask.
> * This causes the `addToken` method on TokensController to get called.
> * This method calls `fetchTokenMetadata`, also on TokensController.
> * This method calls `fetchTokenMetadata` in `assetsUtil`.
> * This function attempts to request information about the token from the Token API. However, because the Token API doesn't support chain 1337, the request fails and an error is thrown.
> * `fetchTokenMetadata` may throw an error if `isTokenListSupportedForNetwork` returns false. If this were to happen, then the TokensController version of `fetchTokenMetadata` would catch this error and return `undefined`. `addToken` expects this to happen and continues on without issue.
> 
> As for the failing extension e2e test — ["Add existing token using search renders the balance for the chosen token"](https://github.com/MetaMask/metamask-extension/blob/642f6d4cf1655edc519de31075d8b14abe0f8e83/test/e2e/tests/add-hide-token.spec.js#L83) — I looked at why this was failing too. This test seems to be trying to simulate production, because if you do remove `GANACHE_CHAIN_ID` from `isTokenListSupportedForNetwork`, then it removes the option to search for tokens from the UI. The way that the extension has gotten around this is with a check in [`import-tokens-modal.js`](https://github.com/MetaMask/metamask-extension/blob/642f6d4cf1655edc519de31075d8b14abe0f8e83/ui/components/multichain/import-tokens-modal/import-tokens-modal.js#L119), which turns it on in e2e tests.
> 
> I see why we make an exception for e2e tests here — we need _some_ way of testing the search box — but I wonder if the way to do this is to use a custom chain ID for Ganache (_not_ 1337, in other words) so that `isTokenListSupportedForNetwork` returns true. Perhaps we will need to extract this test to a different file so that this is easier, I'm not sure.
> 
> To recap, I think we just need to remove `GANACHE_CHAIN_ID` check from `isTokenListSupportedForNetwork` here. That path seems to make the most sense to me, but at the same time I think we should update the e2e test so that it better simulates production instead of localhost.
> 
> What do you think? Would that work?

Thnx @mcmire 🙏  It makes sense to update the e2e test to use mainnet instead of ganache since tokenApi does not support it in the first place; pushed a small update to this and updated the e2e test on extension : https://github.com/MetaMask/metamask-extension/pull/22880

**Review by mcmire — APPROVED** [review]
I hadn't realized you'd already made the change I requested. Sorry for the delay. This looks good to me!
